### ISSUE TITLE
bevy_app re-exports all of bevy_ecs::event

### ISSUE THREAD

**Issue by alice-i-cecile** [bevyengine/bevy#3786]
This is pointless, because `bevy_ecs` is an explicit dependency of all users of `bevy_app`.
Furthermore, it creates confusion in the main Bevy docs, since the `bevy_app` version is listed.

This re-export can and should just be removed.

**Comment by DJMcNab** [comment]
The alternative is to make `bevy_app` export `bevy_ecs` somehow; `bevy_app` is meaningless without `bevy_ecs` anyways.

In some ways, I still think merging them probably makes sense; I can't recall a single usage of `bevy_ecs` without `bevy_app` which hasn't been misguided.

**Comment by CGMossa** [comment]
There are people that want to simply use `bevy_ecs` on its own.
There are efforts put forth to de-couple bevy crates, e.g. #2931 comes to mind also.

**Comment by alice-i-cecile** [comment]
> In some ways, I still think merging them probably makes sense; I can't recall a single usage of bevy_ecs without bevy_app which hasn't been misguided.

Yeah, this is about my perspective too. `bevy_app` is a very thin layer, and even if we roll it into `bevy_ecs` it can still be ignored.

**Comment by CAD97** [comment]
@CGMossa as the author of #2931: there's a difference here.

bevy_app is fundamentally a fairly thin usability layer on top of bevy_ecs, and doesn't add dependencies fundamentally different from bevy_ecs.

bevy_core, however, doesn't have a real "identity" in what it's for. It conflates multiple purposes, with the bevy_time functionality and glue code sort of arbitrarily in one location.

Moving the bevy_core components around as described in #2931 is an improvement to the organization of bevy code because it clarifies the internal dependency graph, and as a side effect, it makes it more practical to depend on a subset of said graph.

bevy_app and bevy_ecs could go either way; either bevy_app doesn't reëxport significant chunks of bevy_ecs, or bevy_ecs absorbs the orchestration of bevy_app. Either way could clean up the clarity of which crate does what, and merging the orchestration into bevy_ecs is simpler.

**Comment by alice-i-cecile** [comment]
> Only because bevy_app is a thin layer around bevy_ecs shouldn't by itself mean there is no value in keeping things specific to the game engine separate (e. g. CoreStage ).

I agree, this and things like `DefaultPlugins` should like in either the root `bevy` module or something like `bevy_core`.

Then, move `App`, `Plugin` and schedule running logic into `bevy_ecs`.

**Comment by bjorn3** [comment]
What about instead moving `Schedule` out of bevy_ecs into bevy_app? This would make bevy_ecs store data, while bevy_app would store app logic.

**Comment by alice-i-cecile** [comment]
> What about instead moving Schedule out of bevy_ecs into bevy_app? This would make bevy_ecs store data, while bevy_app would store app logic.

That's an interesting split. I like it better than our current setup. I wonder if that dividing line would continue to be clean as we add more functionality 🤔

**Comment by hukasu** [comment]
Is this still being pursued?

**Comment by alice-i-cecile** [comment]
It's something that we go back and forth on periodically (including me personally). There's strong arguments in both directions, but nothing to really tip the scales or make it urgent.

**Comment by palapapa** [comment]
To use events in `bevy_ecs`, you have to call `App::add_event`, which only exists in `bevy_app`. According to the comments above, `bevy_app` is just a thin wrapper around `bevy_ecs`. Does that mean that if I want to use events, I have to use `bevy_app` as well?

**Comment by bjorn3** [comment]
`add_event` is implemented as:

```rust
if !self.world.contains_resource::<Events<T>>() {
    EventRegistry::register_event::<T>(self.world_mut());
}
```

which can be replicated without bevy_app.

**Comment by palapapa** [comment]
Is that all I have to do, and events will just work?

**Comment by bjorn3** [comment]
I think so.

**Comment by alice-i-cecile** [comment]
The new generic hierarchy propagation plugin getting shoved into bevy_app would be really nice to clean up with this.
